### PR TITLE
Fix: try harder to resolve `.mjs` files for the browser entries

### LINKED ISSUE #21125: [Bug]: The `Primary` doc block in an attached MDX file shows invalid code preview
### Describe the bug

I'm seeing invalid code previews when using the `Primary` doc block and clicking "Show code":

```jsx
import {
  Meta,
  Primary,
} from "@storybook/blocks";

import * as ButtonStories from "./Button.stories.js";

<Meta of={ButtonStories} />

<Primary />
```

CleanShot 2023-02-16 at 15 52 42@2x

When relying on autodocs, the docs page generated for each component shows the correct JSX in the code preview. But when I add an MDX file and attach that file to a component using the `Meta` block, none of the code previews work (the example above is for the `Primary` block, but I'm also seeing the same error on the `Canvas` block when using `<Canvas of={ButtonStories.Primary} />`).

### To Reproduce

See reproduction repo: https://github.com/elisehein/storybook7-primary-block-code-preview

See README for steps to reproduce, or have a look at the problematic file https://github.com/elisehein/storybook7-primary-block-code-preview/blob/main/src/stories/Button.mdx?plain=1

The repro was created using `npx sb@next init`

### System

```shell
Environment Info:

  System:
    OS: macOS 12.6.3
    CPU: (10) arm64 Apple M1 Pro
  Binaries:
    Node: 16.17.1 - /opt/homebrew/bin/node
    Yarn: 1.22.19 - /opt/homebrew/bin/yarn
    npm: 8.15.0 - /opt/homebrew/bin/npm
  Browsers:
    Chrome: 110.0.5481.96
    Firefox: 109.0.1
    Safari: 15.6.1
  npmPackages:
    @storybook/addon-essentials: ^7.0.0-beta.47 => 7.0.0-beta.47
    @storybook/addon-interactions: ^7.0.0-beta.47 => 7.0.0-beta.47
    @storybook/addon-links: ^7.0.0-beta.47 => 7.0.0-beta.47
    @storybook/blocks: ^7.0.0-beta.47 => 7.0.0-beta.47
    @storybook/react: ^7.0.0-beta.47 => 7.0.0-beta.47
    @storybook/react-webpack5: ^7.0.0-beta.47 => 7.0.0-beta.47
    @storybook/testing-library: ^0.0.14-next.1 => 0.0.14-next.1
```

### Additional context

In my own codebase I was able to fix one particular instance of this by replacing `const Default = {}; export { Default as ComponentName }` in my stories file with `export const Default = {}`. (See [Discord](https://discord.com/channels/486522875931656193/1075399365004361798/1075399365004361798)) I wasn't able to reproduce this here.

### PR DESCRIPTION
**Pull request by ndelangen**
Closes #21125

## What I did

@tmeasday found that the issue was caused by duplicate bundling of addon preview code.

This was because sometimes the `.mjs` file was resolved, and sometimes the `.js` file was resolved.
I've added code to try to find `.mjs` files for resolved files we know will end up in the browser.

I've tested this in the reproduction, and this fixed the issue.

I suspect this will also have quite a positive effect on the bundle-size, and potentially even on build-time.
Good stuff.

### CODE REVIEW

File: code/lib/core-common/src/presets.ts [hunk 1]
```diff
@@ -98,15 +107,21 @@ export const resolveAddonName = (
   // serve it up correctly  when yarn pnp or pnpm is being used.
   // Vite will be broken in such cases, because it does not process absolute paths,
   // and it will try to import from the bare import, breaking in pnp/pnpm.
-  const absolutizeExport = (exportName: string) => {
-    return resolve(`${name}${exportName}`);
+  const absolutizeExport = (exportName: string, preferMJS: boolean) => {
+    const found = resolve(`${name}${exportName}`);
+
+    if (found) {
+      return preferMJS ? resolvePathToMjs(found) : found;
```
Comment (yannbf) [inline code/lib/core-common/src/presets.ts:114]: What's the scenario where mjs would not be preferred?

Comment (ndelangen) [reply code/lib/core-common/src/presets.ts:114]: presets.js in node

Comment (yannbf) [reply code/lib/core-common/src/presets.ts:114]: got it! thanks for explaining

**Review by tmeasday — CHANGES_REQUESTED** [review]
Will this not be a problem if the addon does not have a `.mjs` file in dist? We can't assume that all addons will do that, right?

Are you sure that following through more with the technique of stripping the extension (which previously we were doing for the `managerEntries` but not `previewAnnotations`) isn't the right solution here?

**Comment by ndelangen** [comment]
@tmeasday the better solution would be to maybe use `enhanced-resolve` or something.

I did verify this worked. I was planning on introducing better support for export maps at some point via some method (again `enhanced-resolve` comes to mind). In fact I experimented with it for a bit, but couldn't make it work within the time-box, and so abandoned it in favor of this solution.

**Review by tmeasday — APPROVED** [review]
OK, makes sense then.

**Pull request merged**